feat/go: migrate assets to connector version 0.1.0

NabilHouidi opened the pull request:
## Description
<!-- Describe the goal of the pull request in one or two sentences. -->
- adapt behavior to EDC version 0.1.0 (milestone 9)
- refactor code to simplify
- changes to the assets function signatures (check changes to assetsCrud/main.go)

NabilHouidi commented:
@AhmedGrati @OlfaBensoussia : I tagged @Nazdroth as main reviewer as there were breaking changes but feel free totake a look as well. All feedback welcome 🤗

AhmedGrati commented on go/service/assets/api_op_UpdateAsset_test.go:
```diff
@@ -44,12 +44,12 @@ func Test_UpdateAssetProperties(t *testing.T) {
-	err = apiClient.UpdateAssetProperties(AssetApiInput{
-		AssetProperties: map[string]string{
+	err = apiClient.UpdateAssetProperties(assetId, AssetProperties{
+		PublicProperties: map[string]string{
```
Same as above.

AhmedGrati commented on go/service/assets/api_op_UpdateAsset_test.go:
```diff
@@ -88,13 +88,10 @@ func Test_UpdateAssetDataAddress(t *testing.T) {
+	err = apiClient.UpdateAssetDataAddress(assetId, HttpData{
+		HttpAssetName: httpName,
+		BaseUrl:       httpBaseUrl,
+	},
+	)
```
```suggestion
	})
```

AhmedGrati reviewed (commented):
Nice improvement @NabilHouidi 👍, just some comments below.

NabilHouidi commented on go/internal/connector_api_error.go:
```diff
@@ -26,7 +26,10 @@ func ParseConnectorApiError(response []byte) error {
 		return &wrapError{
 			outer: fmt.Errorf("failed to: %s", ACTION_JSON_UNMARSHAL),
-			inner: err,
+			inner: &wrapError{
+				inner: err,
+				outer: fmt.Errorf("%s", response),
+			},
```
This is so if the error returned by the API is not parsable, we still get the "raw" response in the logs rather than an ambiguous "Failed to parse json"

NabilHouidi commented on go/service/assets/api_op_GetAsset_test.go:
```diff
@@ -41,27 +41,25 @@ func Test_GetAsset(t *testing.T) {
+	assert.Equal(t, asset.PublicProperties["asset:prop:id"], "1234")
 	assert.Equal(t, asset.Id, "1234")
-	assert.Equal(t, asset.CreatedAt, int64(1680172087972))
+	// assert.Equal(t, asset.CreatedAt, int64(1680172087972))
```
I am not sure.
According to openAPI specs, GetAsset should have a `createdAt` field.
When I tested it, it is not returned by the API.

NabilHouidi commented on go/common/sharedtypes/main.go:
```diff
@@ -0,0 +1,17 @@
+package sharedtypes
```
I'll move the `sharedtypes` package under Internal. As it is hidden from the users.  
For the `apivalidator` package, it can be imported and used externally. Go forbids this if the path contains the word `internal`.

NabilHouidi commented:
@AhmedGrati I addressed your comments and fixed most of the unit tests.
Only one unit test fails due to the duplicate `@id` json tag in the `UpdateAssetPropertiesRequestPayload`. cause by [this](https://github.com/eclipse-edc/Connector/issues/3151) bug

The progress is now blocked by https://github.com/eclipse-edc/Connector/issues/3172 and https://github.com/eclipse-edc/Connector/issues/3151

OlfaBensoussia commented on conf/provider-connector.config/addresses.json:
```diff
@@ -3,8 +3,8 @@
     "default": "http://localhost:29191/api",
-    "data": "http://localhost:29193/api/v1/data",
-    "ids": "http://provider-connector:9194/api/v1/ids",
+    "data": "http://localhost:19193/management",
+    "ids": "http://consumer-connector:9194/protocol",
```
@NabilHouidi why did we need to make this change?

NabilHouidi replied:
since v0.1.0, the default addresses of the connector changed.
there's a typo here. line 7 should be `provider-connector`. will fix

OlfaBensoussia commented on conf/consumer-connector.config/addresses.json:
```diff
@@ -3,8 +3,8 @@
     "default": "http://localhost:19191/api",
-    "data": "http://localhost:19193/api/v1/data",
-    "ids": "http://consumer-connector:9194/api/v1/ids",
+    "data": "http://localhost:19193/management",
+    "ids": "http://consumer-connector:9194/protocol",
```
```suggestion
    "management": "http://localhost:19193/management",
    "protocol": "http://consumer-connector:9194/protocol",
```

FYI: this was handled in another `TS` PR, so you can also wait until it's merged and then rebase

OlfaBensoussia commented on go/examples/assetsCrud/main.go:
```diff
@@ -13,7 +13,7 @@ func main() {
 	defaultAddress := "http://localhost:19193/api"
-	managementAddress := "http://localhost:19193/api/v1/data"
+	managementAddress := "http://localhost:19193/management/v2"
 	protocolAddress := "http://localhost:19193/api/v1/ids"
```
```suggestion
	protocolAddress := "http://localhost:19193/protocol"
```

OlfaBensoussia commented on go/service/assets/api_op_CreateAsset_test.go:
```diff
@@ -21,26 +20,31 @@ func Test_CreateAsset(t *testing.T) {
+	"@context": {
+		"edc": "https://w3id.org/edc/v0.0.1/ns/"
+	},
+	"@id": "",
```
is there a reason why we need this here?

NabilHouidi replied:
This was coded when we were blocked by an issue cause by asset having different `id` fields ( `@id` and `asset.@id`).  I believe it is fixed now. Need to retest and check.

OlfaBensoussia commented on go/service/assets/api_op_CreateAsset_test.go:
```diff
@@ -21,26 +20,31 @@ func Test_CreateAsset(t *testing.T) {
 
 		fmt.Fprintf(w, `
 {
 	"createdAt": 1680004526,
```
Since this is a response, it should be
```suggestion
	"edc:createdAt": 1680004526,
```

OlfaBensoussia commented on go/service/assets/api_op_GetAssetProperties_test.go:
```diff
@@ -0,0 +1,56 @@
+	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, `
+{
+	"createdAt": 1680172087972,
```
```suggestion
	"edc:createdAt": 1680172087972,
```

OlfaBensoussia commented on go/service/assets/api_op_UpdateAssetDataAddress_test.go:
```diff
@@ -0,0 +1,54 @@
+}`, string(payload), "invalid payload")
+		assert.Equal(t, "PUT", r.Method, "unexpected method")
+		assert.Equal(t, "/assets/1234/dataaddress", r.URL.Path)
+		w.WriteHeader(204)
```
just to understand what's happening here, why are we returning a success status response code here if we know that the request isn't going through?

NabilHouidi replied:
This is an assert statement. It checks that the received payload is as expected. If it is not, testserver raises an "invalid payload" error. Otherwise, the next statement is executed (another assert).

Once everything is as expected, and no errors were raised, the testserver sends back a `204`.

NabilHouidi commented:
@Nazdroth @AhmedGrati @OlfaBensoussia  I will merge this we can migrate to a separate go repo. we can then fix the `id` issue
